gohcl: Unable to decode block to *hcl.Block field.

I think I have discovered a bug in `gohcl` (I am using v2.11.1). Could you please confirm if this is a legitimate issue, or if I am misusing the package.

The [`gohcl` package docs](https://github.com/hashicorp/hcl/blob/633d41d3df99d6bc4b0fa50d2319344d1efa7a72/gohcl/doc.go#L27) state:

> "block" fields may be of type *hcl.Block or hcl.Body, in which case the corresponding raw value is assigned ... [etc]

However, attempting to decode a block into a field of type `*hcl.Block` produces an error (unless the block is empty).

## Analysis

As best I can tell the issue is with the interplay between these two lines, though there are likely other culprits:

- [decode.go#L266](https://github.com/hashicorp/hcl/blob/633d41d3df99d6bc4b0fa50d2319344d1efa7a72/gohcl/decode.go#L266) in which ` decodeBlockToValue()` performs the check to see if the field can be assigned a `*hcl.Block` value
- [decode.go#L221](https://github.com/hashicorp/hcl/blob/633d41d3df99d6bc4b0fa50d2319344d1efa7a72/gohcl/decode.go#L221) which calls `decodeBlockToValue()` for fields that are pointers.  Specifically, it dereferences the pointer using `v.Elem()`, hence passing a value of type `hcl.Block` _not_ `*hcl.Block`, causing the conditional on line 266 to fail.

The net result is that the field is treated like any other user-defined struct that has no `hcl` field tags, which is why this error only occurs if the block in the `.hcl` file is not empty.

## Reproducing the issue

The following code can be used to reproduce the issue:

```main.go
package main

import (
	"github.com/hashicorp/hcl/v2"
	"github.com/hashicorp/hcl/v2/hclsimple"
)

func main() {
	var output struct {
		TestBlock *hcl.Block `hcl:"my_block,block"`
	}

	err := hclsimple.Decode(
		"reproduce.hcl",
		[]byte(
			`my_block {
				some_attr = true
			}`,
		),
		nil, &output,
	)

	if err != nil {
		panic(err)
	}
}
```

It produces the following output:

```console
❯ go run main.go
panic: reproduce.hcl:2,5-14: Unsupported argument; An argument named "some_attr" is not expected here.

goroutine 1 [running]:
main.main()
	/tmp/hcl-test/main.go:24 +0xf0
exit status 2
```

[This archive](https://github.com/hashicorp/hcl/files/7867144/reproduce-hcl-block-decode-failure.zip) contains the `main.go` file above and `go.mod` / `go.sum` files to help reproduce the issue.

---

Thanks in advance for your help, and for making this fantastic system open source :+1:

--- a/gohcl/decode.go
+++ b/gohcl/decode.go
@@ -258,32 +258,14 @@ func decodeBodyToMap(body hcl.Body, ctx *hcl.EvalContext, v reflect.Value) hcl.D
 }
 
 func decodeBlockToValue(block *hcl.Block, ctx *hcl.EvalContext, v reflect.Value) hcl.Diagnostics {
-	var diags hcl.Diagnostics
+	diags := decodeBodyToValue(block.Body, ctx, v)
 
-	ty := v.Type()
-
-	switch {
-	case blockType.AssignableTo(ty):
-		v.Elem().Set(reflect.ValueOf(block))
-	case bodyType.AssignableTo(ty):
-		v.Elem().Set(reflect.ValueOf(block.Body))
-	case attrsType.AssignableTo(ty):
-		attrs, attrsDiags := block.Body.JustAttributes()
-		if len(attrsDiags) > 0 {
-			diags = append(diags, attrsDiags...)
+	if len(block.Labels) > 0 {
+		blockTags := getFieldTags(v.Type())
+		for li, lv := range block.Labels {
+			lfieldIdx := blockTags.Labels[li].FieldIndex
+			v.Field(lfieldIdx).Set(reflect.ValueOf(lv))
 		}
-		v.Elem().Set(reflect.ValueOf(attrs))
-	default:
-		diags = append(diags, decodeBodyToValue(block.Body, ctx, v)...)
-
-		if len(block.Labels) > 0 {
-			blockTags := getFieldTags(ty)
-			for li, lv := range block.Labels {
-				lfieldIdx := blockTags.Labels[li].FieldIndex
-				v.Field(lfieldIdx).Set(reflect.ValueOf(lv))
-			}
-		}
-
 	}
 
 	return diags

--- a/gohcl/doc.go
+++ b/gohcl/doc.go
@@ -24,11 +24,9 @@
 // expression is assigned, or of any type accepted by gocty, in which case
 // gocty will be used to assign the value to a native Go type.
 //
-// "block" fields may be of type *hcl.Block or hcl.Body, in which case the
-// corresponding raw value is assigned, or may be a struct that recursively
-// uses the same tags. Block fields may also be slices of any of these types,
-// in which case multiple blocks of the corresponding type are decoded into
-// the slice.
+// "block" fields may be a struct that recursively uses the same tags, or a
+// slice of such structs, in which case multiple blocks of the corresponding
+// type are decoded into the slice.
 //
 // "body" can be placed on a single field of type hcl.Body to capture
 // the full hcl.Body that was decoded for a block. This does not allow leftover

